File: android/app/src/main/java/app/farmsy/android/features/discover/SeasonRail.kt
package app.farmsy.android.features.discover

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.wrapContentHeight
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.ShoppingBasket
import androidx.compose.material.icons.filled.WifiOff
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.farmsy.android.LocalTrip
import app.farmsy.android.R
import app.farmsy.android.features.main.LocalShell
import app.farmsy.android.features.main.AppTab
import app.farmsy.android.core.SeasonalItem
import app.farmsy.android.core.Seasons
import app.farmsy.android.core.ShoppingItem
import app.farmsy.android.core.ShoppingItems
import app.farmsy.android.features.whatsnew.SkeletonBox
import app.farmsy.android.ui.theme.EmptyState
import app.farmsy.android.ui.ProductImage
import app.farmsy.android.ui.theme.Badge
import app.farmsy.android.ui.theme.CardShape
import app.farmsy.android.ui.theme.FarmsyColors
import app.farmsy.android.ui.theme.PillButton
import app.farmsy.android.ui.theme.PillSize
import app.farmsy.android.ui.theme.PillVariant
import app.farmsy.android.ui.theme.Radius
import app.farmsy.android.ui.theme.ScreenHeader
import app.farmsy.android.ui.theme.SectionHeader
import app.farmsy.android.ui.theme.Space
import app.farmsy.android.ui.theme.TextRole
import app.farmsy.android.ui.theme.role
import app.farmsy.android.ui.theme.tapCard
import app.farmsy.android.ui.theme.ui
import kotlinx.coroutines.launch
import java.time.Month
import java.time.format.TextStyle
import java.util.Locale

// The year as a vertical rail of twelve months, the current one large and
// green, the past dimmed, the future plain. Nothing is locked: a month is a
// page you can read any time. The Android twin of iOS SeasonRail.swift.

fun monthName(m: Int): String {
    val locale = Locale.getDefault()
    return Month.of(m).getDisplayName(TextStyle.FULL_STANDALONE, locale).replaceFirstChar { it.titlecase(locale) }
}

private fun monthShort(m: Int): String {
    val locale = Locale.getDefault()
    return Month.of(m).getDisplayName(TextStyle.SHORT_STANDALONE, locale).take(3).replaceFirstChar { it.titlecase(locale) }
}

/// One month as a rail node: circle, connector, name, count, up to five thumbs.
@Composable
fun SeasonNode(m: Int, now: Int, items: List<SeasonalItem>, ideas: Int, onOpen: () -> Unit) {
    val current = m == now
    val past = m < now
    val peak = items.filter { it.isPeak(m) }
    val circle = if (current) 56.dp else 44.dp
    Row(
        Modifier.fillMaxWidth().height(IntrinsicSize.Min).tapCard(onTap = onOpen),
        horizontalArrangement = Arrangement.spacedBy(Space.s3),
    ) {
        Column(Modifier.width(56.dp).fillMaxHeight(), horizontalAlignment = Alignment.CenterHorizontally) {
            Box(
                Modifier.size(circle)
                    .background(if (current) FarmsyColors.vivid else if (past) FarmsyColors.creamFill else FarmsyColors.surface, CircleShape)
                    .then(if (!current && !past) Modifier.border(1.dp, FarmsyColors.hairline, CircleShape) else Modifier),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    monthShort(m), style = ui(if (current) 15.sp else 12.sp, FontWeight.Bold),
                    color = if (past) FarmsyColors.inkFaint else FarmsyColors.ink,
                )
            }
            if (m < 12) Box(Modifier.width(3.dp).weight(1f).background(FarmsyColors.hairline))
        }
        Column(Modifier.weight(1f).padding(bottom = Space.s5), verticalArrangement = Arrangement.spacedBy(Space.s2)) {
            Row(horizontalArrangement = Arrangement.spacedBy(Space.s2), verticalAlignment = Alignment.CenterVertically) {
                Text(
                    monthName(m), style = role(if (current) TextRole.HEADING else TextRole.SUBHEADING),
                    color = if (past) FarmsyColors.inkMuted else FarmsyColors.ink,
                )
                if (current) Badge(stringResource(R.string.now_badge), fill = FarmsyColors.vivid, ink = FarmsyColors.ink)
            }
            Text(
                if (items.isEmpty()) stringResource(R.string.season_stored) else stringResource(R.string.season_count_arg, items.size, ideas),
                style = role(TextRole.CAPTION), color = FarmsyColors.inkMuted,
            )
            if (items.isNotEmpty()) {
                Row(Modifier.alpha(if (past) 0.55f else 1f), horizontalArrangement = Arrangement.spacedBy(Space.s2)) {
                    peak.ifEmpty { items }.take(5).forEach { ProductImage(it.imageSlug, it.emoji, 40.dp, corner = 10.dp) }
                }
            }
        }
    }
}

/// One month: what is ripe (peak first), then things to make, each with one
/// button that puts its ingredients on the shopping list. Sheet content.
@Composable
fun MonthSheet(month: Int, onDismiss: () -> Unit) {
    val context = LocalContext.current
    var openProduct by remember { mutableStateOf<SeasonalItem?>(null) }
    val language = remember { ShoppingItems.language(context) }
    val seasonItems by Seasons.items.collectAsState()
    val seasonIdeas by Seasons.ideas.collectAsState()
    val items = remember(seasonItems, month) { Seasons.items(month, seasonItems) }
    val ideas = remember(seasonIdeas, month) { Seasons.ideas(month, seasonIdeas) }

    Column(Modifier.fillMaxSize().background(FarmsyColors.cream)) {
        ScreenHeader(monthName(month), compact = true, onBack = onDismiss)
        Column(
            Modifier.fillMaxWidth().verticalScroll(rememberScrollState()).padding(horizontal = Space.s4).padding(bottom = Space.s8),
            verticalArrangement = Arrangement.spacedBy(Space.s2),
        ) {
            SectionHeader(stringResource(R.string.in_season), top = 0.dp)
            if (items.isEmpty()) {
                Text(stringResource(R.string.season_empty_month), style = role(TextRole.BODY_SM), color = FarmsyColors.inkMuted)
            }
            // A plain 3-column grid: a lazy grid cannot sit inside a scroll column.
            items.chunked(3).forEach { rowItems ->
                Row(Modifier.fillMaxWidth().padding(bottom = Space.s1), horizontalArrangement = Arrangement.spacedBy(Space.s3)) {
                    rowItems.forEach { item ->
                        Column(
                            Modifier.weight(1f).tapCard { openProduct = item },
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(Space.s2),
                        ) {
                            Box(contentAlignment = Alignment.TopEnd) {
                                ProductImage(item.imageSlug, item.emoji, 96.dp, corner = Radius.tile)
                                if (item.isPeak(month)) {
                                    Badge(stringResource(R.string.peak), fill = FarmsyColors.vivid, ink = FarmsyColors.ink, modifier = Modifier.padding(6.dp))
                                }
                            }
                            Text(item.label(language), style = role(TextRole.CAPTION), color = FarmsyColors.ink, maxLines = 1, overflow = TextOverflow.Ellipsis)
                        }
                    }
                    repeat(3 - rowItems.size) { Spacer(Modifier.weight(1f)) }
                }
            }
            if (ideas.isNotEmpty()) {
                SectionHeader(stringResource(R.string.what_to_make))
                CardCarousel(ideas) { idea ->
                    IdeaCard(
                        kicker = null, title = idea.title.text(language), text = idea.body.text(language),
                        image = idea.image, fallback = "🍽️", ingredients = idea.ingredients,
                    )
                }
            }
        }
    }
    openProduct?.let { item ->
        ProductBottomSheet(item.slug, fallbackLabel = item.label(language), fallbackEmoji = item.emoji) { openProduct = null }
    }
}

/// Full-width cards that snap one per page, with dots. The Nime "For you"
/// carousel.
@Composable
fun <T> CardCarousel(items: List<T>, content: @Composable (T) -> Unit) {
    val pager = rememberPagerState { items.size }
    Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(Space.s3)) {
        HorizontalPager(
            state = pager,
            contentPadding = PaddingValues(end = 32.dp),
            pageSpacing = 12.dp,
            verticalAlignment = Alignment.Top,
            modifier = Modifier.fillMaxWidth().wrapContentHeight(),
        ) { i ->
            content(items[i])
        }
        if (items.size > 1) {
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)) {
                items.indices.forEach { i ->
                    Box(Modifier.size(6.dp).background(if (i == pager.currentPage) FarmsyColors.ink else FarmsyColors.hairline, CircleShape))
                }
            }
        }
    }
}

/// Image, kicker, title, body, and one button that puts the ingredients on
/// the shopping list. Shared by the season ideas and the tips.
@Composable
fun IdeaCard(
    kicker: String?, title: String, text: String, image: String, fallback: String, ingredients: List<String>,
    /// Shown when `image` is not bundled (a recipe picture not generated yet).
    fallbackImage: String? = null,
) {
    val context = LocalContext.current
    val trip = LocalTrip.current
    val shell = LocalShell.current
    val language = remember { ShoppingItems.language(context) }
    val wanted by trip.wantedProducts.collectAsState()
    val catalogue by ShoppingItems.items.collectAsState()
    val seasonItems by Seasons.items.collectAsState()
    var added by remember { mutableStateOf(false) }

    // Shopping ids the ingredients resolve to: a shopping id as itself, a
    // seasonal slug through the shopping item that shares its photograph,
    // else a custom item named after it.
    val listIds = remember(ingredients, catalogue, seasonItems) {
        ingredients.map { ing ->
            if (ShoppingItems.item(ing) != null) return@map ing
            val season = seasonItems.firstOrNull { it.slug == ing } ?: return@map ShoppingItem.custom(ing).id
            catalogue.firstOrNull { it.imageSlug == season.imageSlug }?.id ?: ShoppingItem.custom(season.label(language)).id
        }
    }
    val onList = added || (listIds.isNotEmpty() && listIds.all { it in wanted })

    Column(
        Modifier.fillMaxWidth().shadow(8.dp, CardShape).background(FarmsyColors.surface, CardShape).padding(Space.s5),
        verticalArrangement = Arrangement.spacedBy(Space.s3),
    ) {
        Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) { ProductImage(image, fallback, 160.dp, corner = Radius.tile, fallbackSlug = fallbackImage) }
        if (kicker != null) Text(kicker.uppercase(), style = role(TextRole.LABEL), color = FarmsyColors.farmGreen)
        // Lines reserved, not just capped: HorizontalPager is lazy and measures only the
        // realised page(s), so a shorter title/body on another card would otherwise shrink
        // this page and the row would resize while swiping.
        Text(title, style = role(TextRole.HEADING), color = FarmsyColors.ink, minLines = 2, maxLines = 2, overflow = TextOverflow.Ellipsis)
        Text(text, style = role(TextRole.BODY_SM), color = FarmsyColors.inkMuted, minLines = 4, maxLines = 4, overflow = TextOverflow.Ellipsis)
        Spacer(Modifier.weight(1f))
        if (ingredients.isNotEmpty()) {
            PillButton(
                // Once the ingredients are on the list, the useful next step is
                // seeing the list. This used to say "On your list" on a disabled
                // button, which is a dead end at exactly the moment of intent.
                stringResource(if (onList) R.string.view_my_list else R.string.put_on_my_list),
                if (onList) PillVariant.SOFT else PillVariant.PRIMARY, PillSize.SMALL,
                icon = if (onList) Icons.AutoMirrored.Filled.ArrowForward else Icons.Filled.ShoppingBasket,
            ) {
                if (onList) {
                    // showTab closes whatever sheet this card is inside, so the
                    // shopping list is actually visible rather than behind it.
                    shell.showTab(AppTab.SHOPPING)
                } else {
                    listIds.filter { it !in wanted }.forEach { trip.toggleProduct(it) }
                    added = true
                }
            }
        }
    }
}

/// Four skeleton rows while the calendar loads; a retry once the fetch failed.
@Composable
fun SeasonLoading() {
    val failed by Seasons.loadFailed.collectAsState()
    val scope = rememberCoroutineScope()
    if (failed) {
        EmptyState(
            Icons.Filled.WifiOff, stringResource(R.string.calendar_failed), stringResource(R.string.check_connection),
            action = stringResource(R.string.try_again) to { scope.launch { Seasons.loadIfNeeded() } },
        )
    } else {
        repeat(4) { SkeletonBox(cornerRadius = Radius.card, modifier = Modifier.fillMaxWidth().height(84.dp)) }
    }
}

File: Farmsy/Features/Discover/SeasonRail.swift
import SwiftUI

/// The year as a vertical rail of twelve months, the current one large and
/// green, the past dimmed, the future plain. Nothing is locked: a month is a
/// page you can read any time, which is the point of a calendar.
///
/// The rail is Cropsy's timeline (a straight rail, a circle per node, a
/// connector that stretches to the row's height), not a winding path: every
/// node is one tap and the thumb never has to hunt.
struct SeasonRail: View {
    @Binding var openMonth: Int?
    @State private var seasons = Seasons.shared

    private var months: [Int] { Array(1...12) }

    var body: some View {
        if seasons.loadFailed && !seasons.loaded {
            EmptyState(icon: "wifi.slash", title: String(localized: "The calendar didn't load"),
                       text: String(localized: "Check your connection and try again."),
                       action: (String(localized: "Try again"), { Task { await seasons.loadIfNeeded() } }))
        } else if !seasons.loaded {
            ForEach(0..<4, id: \.self) { _ in SkeletonBox(cornerRadius: Radius.card).frame(height: 84) }
        } else {
            Text(String(localized: "What is ripe, month by month, and what to make with it. Tap a month."))
                .role(.bodySm, .inkMuted)
                .padding(.bottom, Space.s2)
            VStack(spacing: 0) {
                ForEach(months, id: \.self) { m in
                    node(m)
                }
            }
        }
    }

    private func node(_ m: Int) -> some View {
        let now = seasons.month
        let state: NodeState = m == now ? .current : (m < now ? .past : .future)
        let items = seasons.items(in: m)
        let peak = items.filter { $0.isPeak(month: m) }
        let ideas = seasons.ideas(in: m)
        return Button {
            Haptics.tap()
            openMonth = m
        } label: {
            HStack(alignment: .top, spacing: Space.s3) {
                // The rail: circle, then a connector that fills the row.
                VStack(spacing: 0) {
                    ZStack {
                        Circle()
                            .fill(state == .current ? Color.vivid : (state == .past ? Color.creamFill : Color.surface))
                            .overlay(Circle().stroke(state == .future ? Color.hairline : Color.clear, lineWidth: 1))
                        Text(monthShort(m))
                            .font(.ui(state == .current ? 15 : 12, .bold))
                            .foregroundStyle(state == .past ? Color.inkFaint : Color.ink)
                    }
                    .frame(width: state == .current ? 56 : 44, height: state == .current ? 56 : 44)
                    if m < 12 {
                        Rectangle().fill(Color.hairline).frame(width: 3).frame(maxHeight: .infinity)
                    }
                }
                .frame(width: 56)

                VStack(alignment: .leading, spacing: Space.s2) {
                    HStack(spacing: Space.s2) {
                        Text(monthName(m)).role(state == .current ? .heading : .subheading,
                                                 state == .past ? .inkMuted : .ink)
                        if state == .current {
                            Badge(text: String(localized: "NOW"), fill: .vivid, ink: .ink)
                        }
                    }
                    Text(items.isEmpty
                         ? String(localized: "Stored crops and the greenhouse.")
                         : String(localized: "\(items.count) in season · \(ideas.count) ideas"))
                        .role(.caption, .inkMuted)
                    if !peak.isEmpty || !items.isEmpty {
                        HStack(spacing: Space.s2) {
                            ForEach((peak.isEmpty ? items : peak).prefix(5)) { item in
                                ProductImage(slug: item.imageSlug, fallback: item.emoji, size: 40, corner: 10)
                            }
                        }
                        .opacity(state == .past ? 0.55 : 1)
                    }
                }
                .padding(.bottom, Space.s5)
                Spacer(minLength: 0)
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .fixedSize(horizontal: false, vertical: true)
    }

    private enum NodeState { case past, current, future }

    static func monthName(_ m: Int) -> String {
        let f = DateFormatter()
        f.locale = Locale.current
        return f.standaloneMonthSymbols[m - 1].capitalized
    }
    private func monthName(_ m: Int) -> String { Self.monthName(m) }
    private func monthShort(_ m: Int) -> String {
        let f = DateFormatter(); f.locale = Locale.current
        return String(f.shortStandaloneMonthSymbols[m - 1].prefix(3)).capitalized
    }
}

/// One month: what is ripe (peak first), then five things to make, each with
/// one button that puts its ingredients on the shopping list.
struct MonthSheet: View {
    let month: Int
    @Environment(\.dismiss) private var dismiss
    @Environment(\.shell) private var shell
    @Environment(FarmsStore.self) private var farms
    @Environment(LocationManager.self) private var locationManager
    @AppStorage("searchRadiusKm") private var radiusKm = 15.0
    @State private var seasons = Seasons.shared
    @State private var openProduct: SeasonalItem?

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(SeasonRail.monthName(month), compact: true, onBack: { dismiss() })
                .padding(.top, Space.s3)
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: Space.s2) {
                    let items = seasons.items(in: month)
                    SectionHeader(title: String(localized: "In season")).padding(.top, 0)
                    if items.isEmpty {
                        Text(String(localized: "Little grows outdoors this month. Farm shops sell what was stored: potatoes, onions, beets, apples, and the cheese and eggs that never stop."))
                            .role(.bodySm, .inkMuted)
                    }
                    LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible()), GridItem(.flexible())], spacing: Space.s3) {
                        ForEach(items) { item in
                            Button {
                                Haptics.tap()
                                openProduct = item
                            } label: {
                                VStack(spacing: Space.s2) {
                                    ZStack(alignment: .topTrailing) {
                                        ProductImage(slug: item.imageSlug, fallback: item.emoji, size: 96, corner: Radius.tile)
                                        if item.isPeak(month: month) {
                                            Badge(text: String(localized: "PEAK"), fill: .vivid, ink: .ink).padding(6)
                                        }
                                    }
                                    Text(item.label).role(.caption).lineLimit(1)
                                }
                            }
                            .buttonStyle(.plain)
                        }
                    }
                    let ideas = seasons.ideas(in: month)
                    if !ideas.isEmpty {
                        SectionHeader(title: String(localized: "What to make with it"))
                        CardCarousel(items: ideas) { idea in
                            IdeaCard(kicker: nil, title: idea.title.text, text: idea.body.text,
                                     image: idea.image, fallback: "🍽️", ingredients: idea.ingredients)
                        }
                    }
                }
                .padding(.horizontal, Space.s4)
                .padding(.bottom, Space.s8)
            }
        }
        .background(Color.cream.ignoresSafeArea())
        .sheet(item: $openProduct) { item in
            ProductSheet(slug: item.slug, fallbackLabel: item.label, fallbackEmoji: item.emoji)
                .presentationDetents([.large])
                .presentationDragIndicator(.visible)
                .presentationCornerRadius(Radius.sheet)
        }
    }
}

/// Full-width cards that snap one per page, with dots. The Nime "For you"
/// carousel.
struct CardCarousel<Item: Identifiable, Content: View>: View {
    let items: [Item]
    @ViewBuilder let content: (Item) -> Content
    @State private var page: Int = 0

    var body: some View {
        VStack(spacing: Space.s3) {
            GeometryReader { geo in
                let w = geo.size.width
                ScrollView(.horizontal, showsIndicators: false) {
                    LazyHStack(spacing: Space.s3) {
                        ForEach(items) { item in
                            content(item).frame(width: w)
                        }
                    }
                    .scrollTargetLayout()
                }
                .scrollTargetBehavior(.viewAligned)
                .scrollPosition(id: Binding(
                    get: { items.indices.contains(page) ? items[page].id : nil },
                    set: { id in if let i = items.firstIndex(where: { $0.id == id }) { page = i } }))
            }
            .frame(height: 380)
            if items.count > 1 {
                HStack(spacing: 6) {
                    ForEach(items.indices, id: \.self) { i in
                        Circle().fill(i == page ? Color.ink : Color.hairline).frame(width: 6, height: 6)
                    }
                }
                .frame(maxWidth: .infinity)
            }
        }
    }
}

/// Image, kicker, title, body, and one button that puts the ingredients on
/// the shopping list. Shared by the season ideas and the tips.
struct IdeaCard: View {
    let kicker: String?
    let title: String
    let text: String
    let image: String
    /// Shown when `image` is not bundled (a recipe picture not generated yet).
    var fallbackImage: String? = nil
    let fallback: String
    let ingredients: [String]

    @Environment(TripStore.self) private var trip
    @Environment(\.shell) private var shell
    @State private var catalogue = ShoppingItems.shared
    @State private var seasons = Seasons.shared
    @State private var added = false

    /// Shopping ids the ingredients resolve to: a shopping id as itself, a
    /// seasonal slug through the shopping item that shares its photograph,
    /// else a custom item named after it.
    private var listIds: [String] {
        ingredients.map { ing in
            if catalogue.item(id: ing) != nil { return ing }
            if let season = seasons.items.first(where: { $0.slug == ing }) {
                if let twin = catalogue.items.first(where: { $0.imageSlug == season.imageSlug }) { return twin.id }
                return ShoppingItem.custom(season.label).id
            }
            return ShoppingItem.custom(ing).id
        }
    }

    private var allOnList: Bool { !listIds.isEmpty && listIds.allSatisfy { trip.wantedProducts.contains($0) } }

    var body: some View {
        VStack(alignment: .leading, spacing: Space.s3) {
            ProductImage(slug: image, fallbackSlug: fallbackImage, fallback: fallback, size: 160, corner: Radius.tile)
                .frame(maxWidth: .infinity)
            if let kicker {
                Text(kicker).role(.label, .farmGreen).textCase(.uppercase)
            }
            Text(title).role(.heading).lineLimit(2).fixedSize(horizontal: false, vertical: true)
            Text(text).role(.bodySm, .inkMuted).lineLimit(4).fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
            if !ingredients.isEmpty {
                Button {
                    Haptics.success()
                    for id in listIds where !trip.wantedProducts.contains(id) { trip.toggleProduct(id) }
                    added = true
                } label: {
                    Label(allOnList || added ? String(localized: "On your list") : String(localized: "Put it on my list"),
                          systemImage: allOnList || added ? "checkmark" : "basket")
                }
                .buttonStyle(PillButtonStyle(allOnList || added ? .soft : .primary, size: .small))
                .disabled(allOnList || added)
            }
        }
        .padding(Space.s5)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(Color.surface, in: RoundedRectangle(cornerRadius: Radius.card, style: .continuous))
    }
}
